domain/relation/state/relation.go

```diff
@@ -481,6 +483,389 @@ WHERE  relation_id = $getRelation.relation_id
+	return nil
+}
+
+// getApplicationOfUnit returns the ID of the application associated with the
+// unit.
+func (st *State) getApplicationOfUnit(
+	ctx context.Context,
+	tx *sqlair.TX,
+	unitUUID unit.UUID,
+) (application.ID, error) {
+	type getUnitApp struct {
+		ApplicationUUID application.ID `db:"application_uuid"`
+		UnitUUID        unit.UUID      `db:"uuid"`
+	}
+	args := getUnitApp{
+		UnitUUID: unitUUID,
+	}
+	stmt, err := st.Prepare(`
+SELECT &getUnitApp.*
+FROM   unit
+WHERE  uuid = $getUnitApp.uuid
+`, args)
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	err = tx.Query(ctx, stmt, args).Get(&args)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return "", relationerrors.UnitNotFound
+	} else if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	return args.ApplicationUUID, nil
+}
+
+// getApplicationsInRelation gets all the applications that are in the given
+// relation.
+func (st *State) getApplicationsInRelation(
+	ctx context.Context,
+	tx *sqlair.TX,
+	uuid corerelation.UUID,
+) ([]application.ID, error) {
+	type applicationUUID struct {
+		UUID application.ID `db:"application_uuid"`
+	}
+	relUUID := relationUUID{
+		UUID: uuid.String(),
+	}
+	stmt, err := st.Prepare(`
+SELECT &applicationUUID.application_uuid
+FROM   relation_endpoint re 
+JOIN   application_endpoint ae ON re.endpoint_uuid = ae.uuid
+WHERE  re.relation_uuid = $relationUUID.uuid
+`, relUUID, applicationUUID{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	var apps []applicationUUID
+	err = tx.Query(ctx, stmt, relUUID).GetAll(&apps)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return nil, relationerrors.RelationNotFound
+	} else if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var ids []application.ID
+	for _, app := range apps {
+		ids = append(ids, app.UUID)
+	}
+
+	return ids, nil
+}
+
+// checkSubordinateUnitCanEnterScope checks that subordinate units can enter
+// scope. Under the following conditions, a subordinate unit cannot enter a
+// relation that is not with its principle application:
+//   - The relation has scope container.
+//   - The unit entering scope is a subordinate.
+//   - The other application is not a subordinate.
+func (st *State) checkSubordinateUnitCanEnterScope(
+	ctx context.Context,
+	tx *sqlair.TX,
+	relUUID corerelation.UUID,
+	unitUUID unit.UUID,
+	otherApplication application.ID,
+) error {
+	// Check that the relation is a container scoped relation.
+	scope, err := st.getRelationScope(ctx, tx, relUUID)
+	if err != nil {
+		return errors.Errorf("getting relation scope: %w", err)
+	}
+	if scope != charm.ScopeContainer {
+		return nil
+	}
+
+	// Check that the other application in the relation is not a subordinate
+	// application, if it is, we have a relation between two subordinates, which
+	// is OK.
+	if subordinate, err := st.isSubordinate(ctx, tx, otherApplication); err != nil {
+		return errors.Errorf("checking if application is subordinate: %w", err)
+	} else if !subordinate {
+		return nil
+	}
```

Comment:
todo: start the method with this check and only get the relation scope if it's a subordinate.